Core/include/Acts/Navigation/SurfaceArrayNavigationPolicy.hpp

```diff
@@ -0,0 +1,77 @@
+// This file is part of the ACTS project.
+//
+// Copyright (C) 2016 CERN for the benefit of the ACTS project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "Acts/Geometry/SurfaceArrayCreator.hpp"
+#include "Acts/Navigation/INavigationPolicy.hpp"
+#include "Acts/Surfaces/SurfaceArray.hpp"
+
+#pragma once
+
+namespace Acts {
+
+/// A navigation policy that internally uses the Gen1 @c SurfaceArray class
+class SurfaceArrayNavigationPolicy : public INavigationPolicy {
+ public:
+  /// Enum for configuring which type of surface array to produce. This affects
+  /// the projection that is used for creating the binning structure.
+  enum class LayerType { Cylinder, Disc, Plane };
+
+  /// Config struct to configure the surface array navigation
+  struct Config {
+    /// The type of the layer
+    LayerType layerType = LayerType::Cylinder;
+    /// The number of bins in the local directions. The interpretation depends
+    /// on the layer type.
+    std::pair<std::size_t, std::size_t> bins;
+  };
+
+  /// Main constructor, which internally creates the surface array acceleration
+  /// structure.
+  /// @note Expects that all relevant surfaces are registered with @p volume.
+  ///       Only selects sensitive surfaces for the surface array.
+  /// @param gctx The geometry context
+  /// @param volume The *layer volume* to construct the surface array from
+  /// @param logger A logging instance
+  /// @param config The configuration for the surface array
+  explicit SurfaceArrayNavigationPolicy(const GeometryContext& gctx,
+                                        const TrackingVolume& volume,
+                                        const Logger& logger, Config config);
+
+  /// Update the navigation state from the surface array
+  /// @param args The navigation arguments
+  void updateState(const NavigationArguments& args) const;
```

Comment:
this interface look a bit suspicious to me because it is called `updateState` and we only provide a `const &`. I wonder if it would be better to separate the arguments a little like: `Stream &, const NavigationArguments &, Logger &`

Reply:
I also think this entry point is way to generic. `updateState` could mean anything: did we enter the volume? did we just finish a step? I would rather argue for concrete interface for each of these conditions.

Reply:
Again, `NavigationArguments` is effectively a keyword argument set, in my mind. I can split it up if that's cleaner, but I was thinking that this allows changing the interface without having to change it in many places, e.g. if we want to add an argument. `Logger` might be a special case, but otherwise I don't think there's a conceptual difference. 

The const ref is to enable passing an rvalue to the call from the outside. It doesn't prevent you from mutating mutable refs inside the structure, e.g. the navigation stream. The position and direction are const in this way however. 

I don't have an opinion on the name. 

Concerning interface and other methods: this one is really just meant as the _main_ navigation callback. I would separate this from an _on enter_ or _on exit_ call, and not try to pipe that through with argument switching. This one would provide navigation candidates, and initiate renavigation inside a volume if needed. We'd have to see if we end up only calling it after volume enter, or more often (after every step?) to enable renavigation. 

Other methods I would just add to the `INavigationPolicy` as virtual functions, and have `MultiNavigationPolicy` broadcast them to its children (which is again a reason why I think this level of composition makes sense).

Reply:
We also lose the ability to reuse memory if we return by value. I don't think returning by value is actually an option at all. 

Also the renavigation will eventually require modifying candidates that already exist, I think. 

I agree with you main point that we shouldn't pipe through the whole state to blanket modification, but this not that, I don't think. I specifically called it `Arguments` to make clear that we pick and choose exactly what we pass in here as explicit arguments, which happen to be contained in a struct. 

Technically, right now, I agree that this is on entry. My thinking is that once we have the volume local state, where a policy can decide not to act on subsequent calls by flipping a bool in the state, we could make the navigation call this more often, with possibly no action being taken. 

For right now, `initializeCandidates` sounds good to me, but as said above we might revisit that in the future.

Reply:
> As long as this is the case I would not generalize the interface.

Just to clarify; that's not the intention, this function is not supposed to be the general and only entry point. Ultimately, I imagine this will be the "hot call" that updates the candidates for example at each step. At that point, "on entry" would be a separate function. 

At this time, this function serves as the "on entry" function, but it will not end up doing both at the same time.

Reply:
> We also lose the ability to reuse memory if we return by value. I don't think returning by value is actually an option at all. Also, I thought the plan was to use the navigation streams that @asalzburger added? I don't think that can work with returned candidates.

Right, I see the problem with the return by value but that does not stop us mentioning the output parameter explicitly in the function definition.

> Also the renavigation will eventually require modifying candidates that already exist, I think.

I don't think that makes sense. Re-navigation == re-initialization which makes all our lives much easier. We kill the candidates vector and create a new one.

---

I think my general problem with the gen2 navigation and this proposal is that we push navigation logic into the delegates but this is not what I would do. I would push the candidate access logic to the delegate and let the navigator do the rest. This gives us a clean separation of who is responsible of what and avoids potential conflicts and re-doing of work.

IMO the navigation policy provides a plain set of candidates which includes surfaces and portals which are intersected, filtered and sorted by the navigator. In this case we do not even have to run intersections in the navigation delegate and we can a plain vector of unintersected candidates the different policies can push to.

In the future, if we want to have re-navigation logic we can provide a way to validate the current position and direction against a provided cone/whatsoever which is checked by the navigator and it will issue a renavigation if necessary.

Reply:
> Re-navigation == re-initialization which makes all our lives much easier. We kill the candidates vector and create a new one.

I think that should be done inside the policy, not by the navigator. Otherwise we go back to Gen1 where the global navigator which implements the entire navigation logic for all geometric configurations. The policy is then literally only the acceleration structure, but we lose the flexibility in the process. 

> In the future, if we want to have re-navigation logic we can provide a way to validate the current position and direction against a provided cone/whatsoever which is checked by the navigator and it will issue a renavigation if necessary.

Like above, I think the cone should be internal to the policy. Only the policy knows the validity cone, where and when to create and restore it etc. 

If we only do top level renavigation from the navigator down, then we're completely back at the Gen1 model, with the caveat that we can't really implement muon system, TRT, other navigation solutions.

Reply:
I don't think we should put navigation code on the geometry. We can allow the user to inject code into the `Navigator` but I would not do that through the geometry. If the user needs different navigation it might be better to just write a custom navigator.

I don't see the navigation for muons or TRT being different to anything else we do. There are surfaces and there are portals. We go through them in a sequence which is well defined and can be implemented in a single place. How to trigger re-navigation can still be specialized through the policy.

Sorry but I just don't see that level of customization being useful. This sounds like we put another Actor chain into the propagation but now it lives inside the navigator and has only partly access to the propagation. But maybe I am just missing the use case you and @asalzburger have in mind.

Reply:
> I don't think we should put navigation code on the geometry. We can allow the user to inject code into the Navigator but I would not do that through the geometry. If the user needs different navigation it might be better to just write a custom navigator.

Really? I think this was the original reason to depart from Gen1 navigation. That we want to put the surface lookup and things like ray navigation **in the geometry**, because it depends on the geometry. 

> I don't see the navigation for muons or TRT being different to anything else we do. There are surfaces and there are portals. We go through them in a sequence which is well defined and can be implemented in a single place. 

It's not, the acceleration structure and lookup pattern fully depends on the geometry. 

> How to trigger re-navigation can still be specialized through the policy.

But the policy needs to then be able to take action. 

> Sorry but I just don't see that level of customization being useful. This sounds like we put another Actor chain into the propagation but now it lives inside the navigator and has only partly access to the propagation. But maybe I am just missing the use case you and @asalzburger have in mind.


I don't see a way we can implement TRT, MS or other non silicon tracker geometries with a god navigator. We tried in Gen1, it became impossibly complicated, we try again. 

If we need to list out all navigation modes (from layers to everything else) we just go all the way back to square one.